src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -76,15 +76,185 @@ public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterIn
       throw new RuntimeException(e);
     }
     log.info("Registered framework frameworkId=" + frameworkId.getValue());
+    liveState.updateReconciliationTimestamp();
+    driver.reconcileTasks(Collections.<Protos.TaskStatus>emptyList());
   }
-
   @Override
   public void reregistered(SchedulerDriver driver, MasterInfo masterInfo) {
-    log.info("Reregistered framework.");
+    log.info("Reregistered framework: starting task reconcile");
+    liveState.updateReconciliationTimestamp();
+    driver.reconcileTasks(Collections.<Protos.TaskStatus> emptyList());
   }
 
+  @Override
+  public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
+    log.info(String.format(
+        "Received status update for taskId=%s state=%s message='%s' stagingTasks.size=%d",
+        status.getTaskId().getValue(),
+        status.getState().toString(),
+        status.getMessage(),
+        liveState.getStagingTasksSize()));
+
+    if (!isStagingState(status)) {
+      liveState.removeStagingTask(status.getTaskId());
+    }
+
+    if (isTerminalState(status)) {
+      liveState.removeRunningTask(status.getTaskId());
+      persistentState.removeTaskId(status.getTaskId().getValue());
+      if (liveState.reconciliationComplete()) {
+        correctCurrentPhase();
+      }
+
+    } else if (isRunningState(status)) {
+      liveState.updateTaskForStatus(status);
+
+      log.info(String.format("Current Acquisition Phase: %s", liveState
+          .getCurrentAcquisitionPhase().toString()));
+
+      switch (liveState.getCurrentAcquisitionPhase()) {
+        case RECONCILING_TASKS :
+          if (liveState.reconciliationComplete()) {
+            reconcilePersistentState();
+            correctCurrentPhase();
+          }
+          break;
+        case JOURNAL_NODES :
+          if (liveState.getJournalNodeSize() == conf.getJournalNodeCount()) {
+            correctCurrentPhase();
+          }
+          break;
+        case NAME_NODE_1 :
+          if (liveState.getNameNodeSize() == (HDFSConstants.TOTAL_NAME_NODES - 1)) {
+            correctCurrentPhase();
+          }
+          break;
+        case NAME_NODE_2 :
+          if (liveState.getNameNodeSize() == HDFSConstants.TOTAL_NAME_NODES) {
+            reloadConfigsOnAllRunningTasks(driver);
+            if (!liveState.isNameNode1Initialized()
+                && !status.getMessage().equals(HDFSConstants.NAME_NODE_INIT_MESSAGE)) {
+              sendMessageTo(
+                  driver,
+                  liveState.getFirstNameNodeTaskId(),
+                  liveState.getFirstNameNodeSlaveId(),
+                  HDFSConstants.NAME_NODE_INIT_MESSAGE);
+            }
+            if (liveState.isNameNode1Initialized()
+                && !liveState.isNameNode2Initialized()) {
```

Comment:
Couldn't we make this an `else if` as in `if(!nn1Init) { send(Init); } else if (!nn2Init) { send(Bootstrap); } else { correctCurrentPhase(); }